gmt/tests/test_surface.py

```diff
@@ -0,0 +1,96 @@
+"""
+Tests for surface
+"""
+import os
+
+import pytest
+import xarray as xr
+
+from .. import surface
+from .. import which
+from ..datasets import load_tut_ship
+from ..exceptions import GMTInvalidInput
+from ..helpers import data_kind
+
+TEST_DATA_DIR = os.path.join(os.path.dirname(__file__), "data")
+TEMP_GRID = os.path.join(TEST_DATA_DIR, "tmp_grid.nc")
+
+
+def test_surface_input_file():
+    """
+    Run surface by passing in a filename
+    """
+    fname = which("@tut_ship.xyz", download="c")
+    output = surface(data=fname, I="5m", R="245/255/20/30")
+    assert isinstance(output, xr.Dataset)
+    return output
+
+
+def test_surface_input_data_array():
+    """
+    Run surface by passing in a numpy array into data
+    """
+    ship_data = load_tut_ship()
+    data = ship_data.values  # convert pandas.DataFrame to numpy.ndarray
+    output = surface(data=data, I="5m", R="245/255/20/30")
+    assert isinstance(output, xr.Dataset)
+    return output
+
+
+def test_surface_input_xyz():
+    """
+    Run surface by passing in x, y, z numpy.ndarrays individually
+    """
+    ship_data = load_tut_ship()
+    output = surface(
+        x=ship_data.x, y=ship_data.y, z=ship_data.z, I="5m", R="245/255/20/30"
+    )
+    assert isinstance(output, xr.Dataset)
+    return output
+
+
+def test_surface_input_xy_no_z():
+    """
+    Run surface by passing in x and y, but no z
+    """
+    ship_data = load_tut_ship()
+    with pytest.raises(GMTInvalidInput):
+        surface(x=ship_data.x, y=ship_data.y, I="5m", R="245/255/20/30")
+
+
+def test_surface_wrong_kind_of_input():
+    """
+    Run surface using grid input that is not file/matrix/vectors
+    """
+    ship_data = load_tut_ship()
+    data = ship_data.z.to_xarray()  # convert pandas.Series to xarray.DataArray
+    assert data_kind(data) == "grid"
+    with pytest.raises(GMTInvalidInput):
+        surface(data=data, I="5m", R="245/255/20/30")
+
+
+def test_surface_g_outfile_param():
+    """
+    Run surface with the -Goutputfile.nc parameter.
+    """
+    ship_data = load_tut_ship()
+    data = ship_data.values  # convert pandas.DataFrame to numpy.ndarray
+    try:
+        output = surface(data=data, I="5m", R="245/255/20/30", G=TEMP_GRID)
+        assert os.path.exists(TEMP_GRID)
+        grid = xr.open_dataset(TEMP_GRID)
+        assert output == grid  # check that original output is same as that from file
```

Comment:
This comparison might not actually work. I'm also not sure I like the fact that you get an output grid when specifying an output file. Maybe we should stick with one or the other. I would suggest not loading the grid if `G` is provided and returning nothing in that case.

Reply:
Yes it's not a particularly well thought out test to be honest. The `-G`outputfile.nc parameter is actually a required argument for [`surface`](https://gmt.soest.hawaii.edu/doc/latest/surface.html), but it's technically optional in this current implementation. I've looked at the other gridding modules and they're not consistent. For example, `-G` is required for [`nearneighbor`](https://gmt.soest.hawaii.edu/doc/latest/nearneighbor.html) but optional for [`triangulate`](https://gmt.soest.hawaii.edu/doc/latest/triangulate.html) and [`greenspline`](https://gmt.soest.hawaii.edu/doc/latest/greenspline.html) (they write to stdout).

I could write a lot of if/then statements to check if the `-G`outputfile.nc parameter is used and return either None or an xarray.Dataset, but I'm not sure if that will be intuitive for other users. A user setting a `-G` parameter will expect `gmt.surface` to output to a file, but would they expect it to return None?